metals/src/main/scala/scala/meta/internal/metals/FileDecoderProvider.scala

```diff
@@ -0,0 +1,277 @@
+package scala.meta.internal.metals
+
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+import java.net.URI
+import java.net.URLDecoder
+import java.nio.charset.StandardCharsets
+import java.nio.file.Paths
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Properties
+import scala.util.Try
+
+import scala.meta.cli.Reporter
+import scala.meta.internal.builds.ShellRunner
+import scala.meta.internal.io.FileIO
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metap.Main
+import scala.meta.internal.mtags.SemanticdbClasspath
+import scala.meta.io.AbsolutePath
+import scala.meta.metap.Format
+import scala.meta.metap.Settings
+
+final class FileDecoderProvider(
+    workspace: AbsolutePath,
+    buildTargets: BuildTargets,
+    userConfig: () => UserConfiguration,
+    shellRunner: ShellRunner,
+    fileSystemSemanticdbs: FileSystemSemanticdbs
+)(implicit ec: ExecutionContext) {
+
+  private case class Finder(findFile: String => Option[AbsolutePath])
+  private case class Decoder(decode: AbsolutePath => Future[Option[String]])
+
+  /**
+   * URI format...
+   * metalsDecodeFile:/fileUrl?decoder=decoderName&keyValueArguments
+   * jar:file:///jarPath/jar-sources.jar!/packagedir/file.java
+   *
+   * Examples...
+   * metalsDecodeFile:/somedir/someFile.java.javap.verbose?decoder=javap
+   * metalsDecodeFile:/somedir/someFile.class.javap?decoder=javap&verbose=false
+   * metalsDecodeFile:/somedir/someFile.scala?decoder=semanticdb
+   * metalsDecodeFile:/somedir/someFile.scala.semanticdb?decoder=semanticdb
+   *
+   * errors get logged and sent back as the text contents
+   */
+  def decodedFileContents(uriAsStr: String): Future[String] = {
```

Comment:
I mentioned this in the TASTy pr as well, but I don't love just returning a `Future[String]` here as it makes it more difficult on the client side to know if you are truly getting the string you want, or just some other random error string. It ends up forcing the client to do odd things with string matching to see what they are getting back. I'd much rather this be a more structured return type so it's easy for clients to immediately know if what they are getting back is valid or not instead of no matter what getting just a `String`.

Reply:
Sure.  Something like below?
```
{
  "output": "Some file text"
}
```
and
```
{
  "error": "Some error message"
}
```
Or do you have an example already that I can work off?

Reply:
Yea it can be pretty simple in the case:

```
interface GoodName {
  error?: String
  body?: String
}
```

Or is someone else has a better interface. But yea, I think this will be nicer for clients to be able to differentiate between these cases.